lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,159 @@
+      workspaces = ApplicationController.firecloud_client.workspaces(project_name)
+      workspaces.each do |workspace|
+        ws_attr = workspace.dig('workspace')
+        ws_name = ws_attr['name']
+        ws_project = ws_attr['namespace']
+        ws_owner = ws_attr['createdBy']
+        existing_study = Study.find_by(firecloud_project: ws_project, firecloud_workspace: ws_name)
+        if existing_study.present?
+          puts "skipping #{ws_project}/#{ws_name} as it belongs to #{existing_study.accession}"
+          next
+        else
+          begin
+            if is_workspace_owner?(workspace)
+              print "deleting #{ws_project}/#{ws_name}... "
+              ApplicationController.firecloud_client.delete_workspace(ws_project, ws_name)
+              puts "#{ws_project}/#{ws_name} successfully deleted"
+            else
+              puts "skipping #{ws_project}/#{ws_name} because it was created by #{ws_owner}"
+            end
+          rescue => e
+            puts "Unable to delete #{ws_project}/#{ws_name} due to: #{e.class.name} - #{e.message}"
+          end
+        end
+      end
+      true
+    end
+  end
+
+  ##
+  # security safeguards to prevent accidental deletes of production/staging workspaces and studies
+  ##
+
+  # ensure commands cannot be run on deployed hosts
+  #
+  # * *raises*
+  #   - (ArgumentError) => if hostname matches DISALLOWED_HOSTS
+  #
+  # * *returns*
+  #   - (TrueClass) => true if hostname passes validation
+  def self.validate_hostname!
+    current_host = Socket.gethostname
+    if current_host =~ DISALLOWED_HOSTS
+      raise ArgumentError.new("#{current_host} is not a permitted host for running StudyCleanupTools methods")
+    else
+      true
+    end
+  end
+
+  # ensure commands cannot be run against protected billing projects for production/staging
+  #
+  # * *raises*
+  #   - (ArgumentError) => if billing project matches DISALLOWED_BILLING_PROJECTS, or is not the configured project for this instance
+  #
+  # * *returns*
+  #   - (TrueClass) => true if billing project passes validation
+  def self.validate_billing_project!(project_name)
+    if !permit_billing_project?(project_name)
+      raise ArgumentError.new("#{project_name} is not a permitted billing project for running StudyCleanupTools methods")
+    else
+      true
+    end
+  end
+
+  # ensure commands cannot be run in production/staging environments
+  # can be run manually in development if allow_dev_env is set to true
+  #
+  # * *raises*
+  #   - (ArgumentError) => if environment is not test/pentest, or is development and allow_dev_env = false
+  #
+  # * *returns*
+  #   - (TrueClass) => true if environment passes validation
+  def self.validate_environment!(allow_dev_env = false)
+    environments = %w(test pentest)
+    allow_dev_env ? environments << 'development' : nil
```

Comment:
A plain `if` would be easier to read here.   I'm not a fan of using ternaries for anything other than assignment.

Reply:
Fair point.